Table header cell has assigned cells- changes in response to the Task Force feedback (#1548)
auto-wcag/auto-wcag: _rules/table-header-cell-has-assigned-cells-d0f69e.md

---
id: d0f69e
name: Table header cell has assigned cell
rule_type: atomic
description: |

>   This rule checks that each table header has assigned at least one non-empty cell.

Reviewer: ```suggestion
  This rule checks that each table header is assigned to at least one non-empty cell.
```

>   This rule checks that each table header has assigned data cells in a table element.

Reviewer: "table element" is not really correct here.

  authors:
    - Audrey Maniez
    - Jey Nandakumar
    - Aron Janecki
---

## Applicability

This rule applies to any [HTML element][] with the [semantic role][] of [rowheader][] or [columnheader][] that is [included in the accessibility tree][] and for which all of the following are true:

> - **in a row** the element is a child of an [accessible object][] with the [semantic role][] of [row][]; and

Author: Agree with the definition of "owned by". As for the accessible object, I don't like the Conflict point for the definition of semantic role. As far as I understand it, `<row role="presentation">` would have a semantic role of row but that would breake the rule. An accessible object with the semantic role of row does not do it for me.

> - **table** the element is an [owned element][] of an [accessible object][] with the [semantic role][] of either [table][] or [grid][] that is [visible][]; and

Author: The definition of "owned by" does not fit very well in here. A `row` can be owned by a `table` or it can be owned by `rowgroup` that is owned by a `table`.  I wanted to catch both.

> - **table** the element is an [owned element][] of an [accessible object][] with the [semantic role][] of either [table][] or [grid][] that is [visible][]; and

Author: What about inclusive descendants?

## Applicability

> This rule applies to any HTML element with the [semantic role][] of [rowheader][] or [columnheader][] within a [table][] or [grid][] element, where the [table][] or [grid][] element is [visible][].

Reviewer: `[table]` and `[grid]` refer to ARIA roles (in the references list). I feel it's not super clear that "a table element" actually means "an element with a role of table".

Also, we normally backticks roles and element (``a [`table`][table] element``).

> This rule applies to any HTML element with the [semantic role][] of [rowheader][] or [columnheader][] within a [table][] or [grid][] element, where the [table][] or [grid][] element is [visible][].

Reviewer: I am not sure that "within" is unambiguous enough.
I understand that this is trying to address Wilco's first comment of https://www.w3.org/2002/09/wbs/93339/ACTTableHeaders/results#xQ8 but I think that "within" can easily be understood as "descendant of" (this is the meaning we use for ["inside a web page"](https://act-rules.github.io/glossary/#web-page-html), for example).

I agree with the initial comment, but I'm afraid we need to be more precise in explaining what it means for a cell to be part of a table. (and this is somewhat more complex due to the use of roles, so we cannot fully rely on HTML definitions from https://html.spec.whatwg.org/multipage/tables.html#table-processing-model such as "the element corresponds to a cell [=concept, not `td`/`th`] in the table [=concept, not the `table` element]" 😖

Author: Something in the lines of "within an element with the semantic role of `table` or `grid` that is its closest ancestor in the accessibility tree"- the ancestor in the accessibility tree does not seem correct because it's used in the DOM standard so have a feeling it should not be used for accessibility tree, but please correct me on that.

Reviewer: The problem with "closest ancestor" is that it is the current version and not really working…

Why I understand is that in
```
<table>
<div><div><tr><th>Foo</th></tr></tr>
</table>
```
the `th` is not really "part" of the `table` because the `div` are skipped. This actually gets more complex due to the fact that HTML parsing actually removes the `div` when building the DOM in that case… But we can reach a similar result with roles:
```html
<div role="table">
<div><div>
  <div role="tablerow"><div role="columnheader">Foo</div></div>
</div></div>
</div>
```
But then in that case it's not even clear what "being part of the table" means since the algorithm for forming a table only works on HTML tables, not on ARIA ones…

Using "within" is, I think, not good because of this:
```html
<table>
<tr><td>
  <table hidden>
  <tr><th>Foo</th><tr>
  </table>
</td></tr>
</table>
```
In which case, the `th` is arguably "within" both `table` elements (as in "within the subtree rooted at that element"), but we want to exclude it from the rule, so we need a definition of "within a table"… And copying the table building algorithm doesn't seem smart…

## Expectation 1

> Each target element is [assigned][] to at least one element with a [semantic role][] of [cell][]. The test target and the [cell][] are within an element with the [semantic role][] of [table][] or [grid][].

Reviewer: The second sentence looks like a different expectation.
Also, since the target is part of the table, maybe
```suggestion
Each target element is [assigned][] to at least one element with a [semantic role][] of [cell][]. which is within the same element with the [semantic role][] of [table][] or [grid][].
```
(+ same caveat about "within")

Reviewer: Ah no, I get it… (reading the next expectation…)
So more like
```suggestion
For each test target, either it is not part of a [table][]; or it is [assigned][] to at least one element with a [semantic role][] of [cell][] which is within the sameelement with the [semantic role][] of [table][].
```

Author: Not sure why "either it is not part of a [table][]"? Elements with the semantic role of `rowheader` or `columnheader` have required context roles that eventually end on an element with the semantic role of either `grid` or `table`.

Reviewer: You need the "either" bit because of the table/grid duality:
```html
<table role="grid">
<tr><th role=columnheader>Foo</th></tr>
<tr><td role="gridcell">Bar</td></tr>
</table>
```
Now, this `th` is not assigned to any element with a role of `cell` and thus fails the expectation. Yet, it is good because it is inside a `grid`.

When there are several expectations, a target needs to pass **all** of them in order to pass the rule.

Author: I think that I've addressed that with the current format of the expectations section. The second expectation should account for the `table`/`grid` duality.

## Expectation 2

> Each target element is [assigned][] to at least one element with a [semantic role][] of [gridcell][]. The test target and the [gridcell][] are within an element with the [semantic role][] of [grid][].

Reviewer: Same concerns as for 1st expectation.

## Assumptions

> This rule assumes that table header cells have a relationship conveyed through presentation with data cells within the same table. This excludes edge cases, where there are multiple headers and no data cells. In such scenarios the rule fails, but [success criterion 1.3.1 Info and Relationships][sc1.3.1] could still be satisfied.

Reviewer: I agree with https://github.com/act-rules/act-rules.github.io/issues/1545#issuecomment-777604045 that we could move part of this Assumption to Applicability by requiring that the table/grid contains at least one non-empty cell/gridcell.

Author: What if "table header has assigned at least one element with the semantic role of `cell` that has non-empty accessible name"? 
There might be a non-empty data cell but if it has `aria-hidden` than that would also be a failure under 1.3.1.

Author: If we apply this rule to either a `rowheader` or a `columnheader` and either of those has assigned at least one non-empty element with a semantic role of `cell` or inheriting from cell than that should work. What do you think?

Reviewer: What we need to avoid here is
```html
<table>
<tr><th>Foo</th></tr>
</table>
```
Where the `th` is not assigned to any `cell`, but this is OK because there is no relationship conveyed through presentation, so 1.3.1 doesn't care. But this does fail the rule because the `th` is assigned to no cell.
Rather than having it in an assumption (actually, in addition to the assumption which covers other corner cases), we could have that bit in the Applicability:
"the `table` contains at least one data cell".
This would remove from the start a large chunk of the corner cases covered by this assumption.

## Applicability

> This rule applies to any HTML element with the [semantic role][] of [`rowheader`][] or [`columnheader`][] that is within an element with the [semantic role][] of either [`table`][] or [`grid`][]. The [`table`][] or [`grid`][] is [visible][].

Reviewer: The square brackets should be links.

#### Passed Example 5

Each of the 4 `th` elements has an assigned `td` element because the value of the `headers` attribute on `td` elements references the value of the `id` attribute on the `th` elements.

```html
<table>
	<tr>
		<th id="projects" rowspan="2">Projects</th>
		<th id="objective" colspan="2">Objective</th>
	</tr>
	<tr>
		<th id="1" headers="objective">1</th>
		<th id="2" headers="objective">2</th>
	</tr>
	<tr>
		<td headers="projects">40%</td>
		<td headers="objective 1">20%</td>

> 		<td headers="objective 2">25%</td>

Reviewer: Is it on purpose that the percentages add up to only 95% 🤔

</table>
```

#### Passed Example 7

Each of the 2 `div` elements has an assigned `gridcell` within the same `div` element having a [semantic role][] of `grid`.

```html
<div role="grid">
	<div role="row">
		<div role="columnheader">Room</div>
		<div role="columnheader">Occupants</div>
	</div>
	<div role="row">
		<div role="gridcell"><button>1A</button></div>
		<div role="gridcell"><input type="number" /></div>
	</div>
	<div role="row">
		<div role="gridcell"><button>2A</button></div>
		<div role="gridcell"><input type="number" /></div>
	</div>
</div>
```

### Failed

#### Failed Example 1

> The `th` elements do not have assigned data cells as per the [internal algorithm for scanning and assigning header cells](https://html.spec.whatwg.org/multipage/tables.html#internal-algorithm-for-scanning-and-assigning-header-cells). Their `scope` attribute is in the auto state and there is a non-empty table data slot in the same line.

Reviewer: ```suggestion
The `th` elements do not have assigned data cells as per the [internal algorithm for scanning and assigning header cells](https://html.spec.whatwg.org/multipage/tables.html#internal-algorithm-for-scanning-and-assigning-header-cells). Their `scope` [attribute value][] is `auto` state and there is a non-empty table data slot in the same column or row.
```

		<div role="columnheader">Occupant</div>
	</div>
	<div role="row">

> 		<div role="gridcell"><button>1A</button></div>

Reviewer: I'm not sure why we need to add `button` here (and in the next row).

Author: This was not meant to be a failed example.

</div>
```

> #### Inapplicable Example 7

Reviewer: I think we can keep the example but update the description to point out that the `table` element does not have a role of `table`, so the rule does not apply.
